File: lintcode/python/1274_find_k_pairs_with_smallest_sums.py
#!/usr/bin/python -t

# heap
# heap: every time calling heapq.heappop(), use heapq.heappush() to add next valid one
# time O(klonk)



import heapq

class Solution:
    """
    @param nums1: List[int]
    @param nums2: List[int]
    @param k: an integer
    @return: return List[List[int]]
    """
    def kSmallestPairs(self, nums1, nums2, k):
        # write your code here
        n1 = len(nums1)
        n2 = len(nums2)
        heap = [(nums1[0]+nums2[0], 0, 0)]
        seen = {(0, 0)}
        ret = []
        
        for i in range(min(k, n1*n2)):
            _, index1, index2 = heapq.heappop(heap)
            ret.append([nums1[index1], nums2[index2]])
            
            if index1 + 1 < n1 and (index1+1, index2) not in seen:
                heapq.heappush(heap, (nums1[index1+1]+nums2[index2], index1+1, index2))
                seen.add((index1+1, index2))
            if index2 + 1 < n2 and (index1, index2+1) not in seen:
                heapq.heappush(heap, (nums1[index1]+nums2[index2+1], index1, index2+1))
                seen.add((index1, index2+1))
                
        return ret
        


# heap
# 简单heap方法，将n1+n2, n1, n2 无脑放入最小堆，最后从堆中pop出k个即可
# time O(mn) space O(m+n)


import heapq

class Solution:
    """
    @param nums1: List[int]
    @param nums2: List[int]
    @param k: an integer
    @return: return List[List[int]]
    """
    def kSmallestPairs(self, nums1, nums2, k):
        # write your code here
        heap = []
        
        ret = []
        
        for num1 in nums1:
            for num2 in nums2:
                heapq.heappush(heap, (num1+num2, num1, num2))
                
        for _ in range(min(k, len(heap))):
            _, num1, num2 = heapq.heappop(heap)
            ret.append([num1, num2])
            
        return ret
        

File: lintcode/golang/1274_find_k_pairs_with_smallest_sums.go
//heap
// not ac, just ordering issue.


import (
	"container/heap"
    //"sort"
)

type Node struct {
	val   int
	r int
    c int
}

// An IntHeap is a min-heap of ints.
type NodeHeap []*Node

func (h NodeHeap) Len() int           { return len(h) }
func (h NodeHeap) Less(i, j int) bool { return h[i].val < h[j].val }
func (h NodeHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *NodeHeap) Push(x interface{}) {
	// Push and Pop use pointer receivers because they modify the slice's length,
	// not just its contents.
	*h = append(*h, x.(*Node))
}

func (h *NodeHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func newNode(val, r, c int) *Node {
	return &Node{
		val:   val,
		r: r,
        c: c,
	}
}

/**
 * @param nums1: List[int]
 * @param nums2: List[int]
 * @param k: an integer
 * @return: return List[List[int]]
 */
func kSmallestPairs (nums1 []int, nums2 []int, k int) [][]int {
    // write your code here
	m := len(nums1)
	n := len(nums2)
	ret := make([][]int, 0)
	v := make(map[int]bool)

	minHeap := &NodeHeap{}
	node := &Node{
		val: nums1[0] + nums2[0],
		r:   0,
		c:   0,
	}
	heap.Push(minHeap, node)
	v[0] = true

	for i := 0; i < min(k, m*n); i++ {
		node := heap.Pop(minHeap).(*Node)
		s := []int{nums1[(*node).r], nums2[(*node).c]}
		//sort.Ints(s)
		ret = append(ret, s)
		/*
			if nums1[(*node).r] < nums2[(*node).c] {
				ret = append(ret, []int{nums1[(*node).r], nums2[(*node).c]})
			} else {
				ret = append(ret, []int{nums2[(*node).c], nums1[(*node).r]})
			}
		*/

		if (*node).c+1 < n {
			index := ((*node).r)*n + ((*node).c + 1)
			if _, ok := v[index]; !ok {
				heap.Push(minHeap, &Node{
					val: nums1[(*node).r] + nums2[(*node).c+1],
					r:   (*node).r,
					c:   (*node).c + 1,
				})
				v[index] = true
			}
		}

		if (*node).r+1 < m {
			index := ((*node).r+1)*n + ((*node).c)
			if _, ok := v[index]; !ok {
				heap.Push(minHeap, &Node{
					val: nums1[(*node).r+1] + nums2[(*node).c],
					r:   (*node).r + 1,
					c:   (*node).c,
				})
				v[index] = true
			}
		}

	}

	return ret
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}
